lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+    # @option i12n_hash [Hash{PType => Hash}] 'annotations' Annotations hash. Default is `nil`.
+    # @api public
+    def initialize(name, i12n_hash)
+      @name = name
+      @type = i12n_hash[KEY_TYPE]
+      @override = i12n_hash[KEY_OVERRIDE]
+      @override = false if @override.nil?
+      @final = i12n_hash[KEY_FINAL]
+      @final = false if @final.nil?
+      @annotations = i12n_hash[KEY_ANNOTATIONS]
+      @annotations.freeze unless @annotations.nil?
+    end
+
+    # Delegates to the contained type
+    # @param visitor [TypeAcceptor] the visitor
+    # @param guard [RecursionGuard] guard against recursion. Only used by internal calls
+    # @api public
+    def accept(visitor, guard)
+      @type.accept(visitor, guard)
+      @annotations.each_key { |key| key.accept(visitor, guard) } unless @annotations.nil?
+    end
+
+    # Checks if the this _member_ overrides an inherited member, and if so, that this member is declared with override = true and that
+    # the inherited member accepts to be overridden by this member.
+    #
+    # @param parent_members [Hash{String=>PAnnotatedMember}] the hash of inherited members
+    # @return [PAnnotatedMember] this instance
+    # @raises [Puppet::ParseError] if the assertion fails
+    # @api private
+    def assert_override(parent_members)
+      parent_member = parent_members[@name]
+      if parent_member.nil?
+        raise Puppet::ParseError, "expected #{label} to override an inherited #{feature_type}" if @override
+        self
+      else
+        parent_member.assert_can_be_overridden(self)
+      end
+    end
+
+    # Checks if the given _member_ can override this member.
+    #
+    # @param member [PAnnotatedMember] the overriding member
+    # @return [PAnnotatedMember] its argument
+    # @raises [Puppet::ParseError] if the assertion fails
+    # @api private
+    def assert_can_be_overridden(member)
+      raise Puppet::ParseError, "#{member.label} overrides inherited #{feature_type}" unless self.class == member.class
+      raise Puppet::ParseError, "attempt to override final #{label}" if @final
+      raise Puppet::ParseError, "#{label} attempts to override without having override defined" unless member.override?
+      raise Puppet::ParseError, "attempt to override #{label} with a type that does not match" unless @type.assignable?(member.type)
+      member
+    end
+
+    # @return [Boolean] `true` if this feature cannot be overridden
+    # @api public
+    def final?
+      @final
+    end
+
+    # @return [Boolean] `true` if this feature must override an inherited feature
+    # @api public
+    def override?
+      @override
+    end
+
+    # @api public
+    def hash
+      @name.hash ^ @type.hash
+    end
+
+    # @api public
+    def eql?(o)
+      self.class == o.class && @name == o.name && @type == o.type && @override == o.override? && @final == o.final? && @annotations == o.annotations
```

Comment:
I wonder if annotations should be included - my spontaneous reaction is that they should not. From the perspective of the object they do not really exist... (although they do technically as they need to serialized with the object, and adapters need to find their respective data).